### PR TITLE
#628: custom directives on types

### PR DESCRIPTION
**Pull request by t1**
plus some power-annotations cleanup. I had worked on an ugly change that turned out to be unnecessary, so I removed it, but left the refactoring, as it was good.

I also fixed a lot of warnings that my IDE complained about.

### CODE REVIEW

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java [hunk 1]
```diff
@@ -153,9 +155,7 @@ public static Annotations getAnnotationsForArgument(MethodInfo methodInfo, short
 
         final org.jboss.jandex.Type parameterType = methodInfo.parameters().get(pos);
 
-        Map<DotName, AnnotationInstance> annotationMap = new HashMap<>();
-
-        annotationMap.putAll(getAnnotations(parameterType));
+        Map<DotName, AnnotationInstance> annotationMap = getAnnotations(parameterType);
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java:158]: simplification (suggested by my IDE)

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java [hunk 2]
```diff
@@ -299,23 +299,11 @@ public String toString() {
...
 
     private boolean hasValidMethodAnnotation(DotName annotation) {
-        if (containsKeyAndValidValue(annotation)) {
-            AnnotationInstance annotationInstance = getAnnotation(annotation);
-            if (isMethodAnnotation(annotationInstance)) {
-                return true;
-            }
-        }
-        return false;
+        return containsKeyAndValidValue(annotation) && isMethodAnnotation(getAnnotation(annotation));
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java:302]: simplification (suggested by my IDE)

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java [hunk 3]
```diff
@@ -339,7 +327,7 @@ private AnnotationInstance getAnnotation(DotName key) {
     }
 
     private Optional<String> getStringValue(AnnotationValue annotationValue) {
-        String value = null;
+        String value;
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java:330]: simplification (suggested by my IDE)

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java [hunk 4]
```diff
@@ -367,7 +355,7 @@ private static Annotations getAnnotationsForInputField(FieldInfo fieldInfo, Meth
         }
         if (methodInfo != null) {
             List<org.jboss.jandex.Type> parameters = methodInfo.parameters();
-            if (parameters != null && !parameters.isEmpty()) {
+            if (!parameters.isEmpty()) {
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/Annotations.java:358]: simplification (suggested by my IDE)

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/SchemaBuilder.java [hunk 5]
```diff
@@ -137,34 +152,34 @@ private void addTypesToSchema(Schema schema) {
     }
 
     private void addOutstandingTypesToSchema(Schema schema) {
-        boolean allDone = true;
+        boolean keepGoing = false;
+
         // See if there is any inputs we missed
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/SchemaBuilder.java:157]: I found all the negations very difficult to follow

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/creator/FieldCreator.java [hunk 6]
```diff
@@ -36,59 +36,64 @@ public FieldCreator(ReferenceCreator referenceCreator) {
...
      * @return a Field model object
      */
     public Optional<Field> createFieldForInterface(MethodInfo methodInfo, Reference parentObjectReference) {
         Annotations annotationsForMethod = Annotations.getAnnotationsForInterfaceField(methodInfo);
 
-        if (isGraphQlField(Direction.OUT, null, methodInfo)) {
-            Type returnType = methodInfo.returnType();
+        if (!isGraphQlField(Direction.OUT, null, methodInfo)) {
+            return Optional.empty();
+        }
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/creator/FieldCreator.java:49]: simplification

File: common/schema-builder/src/main/java/io/smallrye/graphql/schema/creator/type/TypeCreator.java [hunk 7]
```diff
@@ -89,10 +108,8 @@ private void addFields(Type type, ClassInfo classInfo, Reference reference) {
         for (ClassInfo c = classInfo; c != null; c = ScanningContext.getIndex().getClassByName(c.superName())) {
             if (InterfaceCreator.canAddInterfaceIntoScheme(c.toString())) { // Not java objects
                 allMethods.addAll(c.methods());
-                if (c.fields() != null && !c.fields().isEmpty()) {
-                    for (final FieldInfo fieldInfo : c.fields()) {
-                        allFields.putIfAbsent(fieldInfo.name(), fieldInfo);
-                    }
+                for (FieldInfo fieldInfo : c.fields()) {
```
Comment (t1) [inline common/schema-builder/src/main/java/io/smallrye/graphql/schema/creator/type/TypeCreator.java:111]: simplification

File: common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Field.java [hunk 8]
```diff
@@ -60,11 +60,10 @@
     public Field() {
     }
 
-    public Field(String methodName, String propertyName, String name, String description, Reference reference) {
+    public Field(String methodName, String propertyName, String name, Reference reference) {
         this.methodName = methodName;
         this.propertyName = propertyName;
         this.name = name;
-        this.description = description;
```
Comment (t1) [inline common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Field.java:67]: as mentioned above, the description is set later

File: common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Operation.java [hunk 9]
```diff
@@ -37,9 +37,9 @@
     public Operation() {
     }
 
-    public Operation(String className, String methodName, String propertyName, String name, String description,
-            Reference reference, final OperationType operationType) {
-        super(methodName, propertyName, name, description, reference);
+    public Operation(String className, String methodName, String propertyName, String name, Reference reference,
+            final OperationType operationType) {
+        super(methodName, propertyName, name, reference);
```
Comment (t1) [inline common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Operation.java:42]: as mentioned above, the description is set later

File: common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java [hunk 10]
```diff
@@ -23,12 +23,13 @@
...
     private Map<String, InputType> inputs = new HashMap<>();
-    private Map<String, Type> types = new HashMap();
-    private Map<String, InterfaceType> interfaces = new HashMap();
-    private Map<String, EnumType> enums = new HashMap();
+    private Map<String, Type> types = new HashMap<>();
+    private Map<String, InterfaceType> interfaces = new HashMap<>();
+    private Map<String, EnumType> enums = new HashMap<>();
 
-    private Map<String, ErrorInfo> errors = new HashMap();
+    private Map<String, ErrorInfo> errors = new HashMap<>();
```
Comment (t1) [inline common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java:32]: typesafety

File: power-annotations/common/src/main/java/com/github/t1/powerannotations/common/Jandex.java [hunk 11]
```diff
@@ -53,6 +54,7 @@ ClassInfo getClassInfo(DotName className) {
 
     @SuppressWarnings("deprecation")
     private static org.jboss.jandex.ClassInfo mock(DotName name) {
+        log.info("create mock for " + name);
```
Comment (t1) [inline power-annotations/common/src/main/java/com/github/t1/powerannotations/common/Jandex.java:57]: This is why I changed all the log handling below

File: power-annotations/utils/jandex/src/main/java/com/github/t1/powerannotations/utils/jandex/JandexAnnotationsLoader.java [hunk 12]
```diff
@@ -97,10 +96,10 @@ public Annotations onField(Class<?> type, String fieldName) {
     public Annotations onMethod(Class<?> type, String methodName, Class<?>... argTypes) {
         ClassInfo classInfo = getClassInfo(type);
         Type[] jandexArgTypes = Stream.of(argTypes).map(Class::getName).map(JandexAnnotationsLoader::createClassType)
-                .collect(toArray(Type.class));
+                .toArray(Type[]::new);
         MethodInfo methodInfo = classInfo.method(methodName, jandexArgTypes);
         if (methodInfo == null) {
-            String[] argTypeNames = Stream.of(argTypes).map(Class::getName).collect(toArray(String.class));
+            String[] argTypeNames = Stream.of(argTypes).map(Class::getName).toArray(String[]::new);
```
Comment (t1) [inline power-annotations/utils/jandex/src/main/java/com/github/t1/powerannotations/utils/jandex/JandexAnnotationsLoader.java:102]: I just learned that this is how to collect into an array

File: power-annotations/maven-plugin/src/main/java/com/github/t1/powerjandex/PowerJandexMojo.java [hunk 13]
```diff
@@ -52,36 +61,45 @@ private Path baseDir() {
...
-                }
-            });
-            return indexer.complete();
-        } catch (IOException e) {
-            throw new RuntimeException("failed to index", e);
+                });
+            } catch (IOException e) {
+                throw new RuntimeException("failed to index " + path, e);
+            }
+        }
```
Comment (t1) [inline power-annotations/maven-plugin/src/main/java/com/github/t1/powerjandex/PowerJandexMojo.java:85]: the nested Scanner class makes it easier to scan other things, too.

File: power-annotations/scanner/src/main/java/com/github/t1/powerannotations/scanner/IndexBuilder.java [hunk 14]
```diff
@@ -33,4 +35,10 @@ private static ClassLoader getClassLoader() {
         ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
         return (classLoader == null) ? ClassLoader.getSystemClassLoader() : classLoader;
     }
+
+    private static Index scan() {
+        Index index = new Scanner().scanClassPath();
+        new PowerAnnotations(index).resolveAnnotations();
+        return index;
+    }
```
Comment (t1) [inline power-annotations/scanner/src/main/java/com/github/t1/powerannotations/scanner/IndexBuilder.java:43]: moved this here from the Scanner class

**Review by phillip-kruger — APPROVED** [review]
This was quite difficult to review, as it seems to contain multiple changes. I assumed the power annotation changes has got nothing to do with the directive feature ?

**Comment by phillip-kruger** [comment]
Thanks @t1 - This feature should allow us to start adding to built in directives. This is awesome.

**Comment by phillip-kruger** [comment]
@t1 Would you mind to backport this to 1.0.x ? At least the directive support.

**Comment by t1** [comment]
> difficult to review

Yes, sorry... I will make separate PRs next time.

> backport this to 1.0.x

I hadn't planed for it. It's also not feature complete, yet. Do you have a broad guess when Quarkus can update to 1.1.x?

**Comment by phillip-kruger** [comment]
I think in 2 months or so

**Pull request merged**